src/Sylius/Bundle/CartBundle/Provider/CartProvider.php

```diff
@@ -127,36 +121,36 @@ public function setCart(CartInterface $cart)
      */
     public function abandonCart()
     {
-        if (null !== $this->cart) {
-            $this->eventDispatcher->dispatch(SyliusCartEvents::CART_ABANDON, new CartEvent($this->cart));
-        }
+        $cart = $this->provideCart();
+
+        $this->eventDispatcher->dispatch(
+            SyliusCartEvents::CART_ABANDON,
+            new GenericEvent($cart)
+        );
 
-        $this->cart = null;
         $this->context->resetCurrentCartIdentifier();
     }
 
     /**
-     * Gets cart by cart identifier.
-     *
-     * @param mixed $identifier
+     * Tries to initialize cart if there is data in storage.
+     * If not returns new instance from resourceFactory
      *
-     * @return CartInterface|null
+     * @return CartInterface
      */
-    protected function getCartByIdentifier($identifier)
+    private function provideCart()
     {
-        return $this->repository->find($identifier);
-    }
+        $cartIdentifier = $this->context->getCurrentCartIdentifier();
+        if ($cartIdentifier) {
```

Comment:
I am not sure about that, but I suppose it is possible to have `$cartIdenetifier` with value 0. In this case we would skip this statement. Nevertheless I think it is better to compare it directly with null, or any other value.